smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TopologicalIndex.java

```diff
@@ -0,0 +1,192 @@
+ * paths from the shape back to itself. This list can be useful for code
+ * generation to generate different code based on if a recursive path
+ * passes through particular types of shapes.
+ */
+public final class TopologicalIndex implements KnowledgeIndex {
+
+    private final Set<Shape> shapes = new LinkedHashSet<>();
+    private final Map<Shape, List<PathFinder.Path>> recursiveShapes = new LinkedHashMap<>();
+
+    public TopologicalIndex(Model model) {
+        // A reverse-topological sort can't be performed on recursive shapes,
+        // so instead, recursive shapes are explored first and removed from
+        // the topological sort.
+        computeRecursiveShapes(model);
+
+        // Next, the model is explored using a DFS so that targets of shapes
+        // are ordered before the shape itself.
+        NeighborProvider provider = NeighborProviderIndex.of(model).getProvider();
+        model.shapes()
+                // Note that while we do not scan the prelude here, shapes from
+                // the prelude are pull into the ordered result if referenced.
+                .filter(FunctionalUtils.not(Prelude::isPreludeShape))
+                .filter(shape -> !recursiveShapes.containsKey(shape))
+                // Sort here to provide a deterministic result.
+                .sorted()
+                .forEach(shape -> visitShape(provider, shape));
+    }
+
+    private void computeRecursiveShapes(Model model) {
+        // PathFinder is used to find all paths from U -> U.
+        PathFinder finder = PathFinder.create(model);
+
+        // The order of recursive shapes is first by the number of edges
+        // (the degree of recursion), and then alphabetically by shape ID.
+        Map<Integer, Map<Shape, List<PathFinder.Path>>> edgesToShapePaths = new TreeMap<>();
+        for (Shape shape : model.toSet()) {
+            if (!Prelude.isPreludeShape(shape) && !(shape instanceof SimpleShape)) {
+                // Find all paths from the shape back to itself.
+                List<PathFinder.Path> paths = finder.search(shape, shape);
+                if (!paths.isEmpty()) {
+                    int edgeCount = 0;
+                    for (PathFinder.Path path : paths) {
+                        edgeCount += path.size();
+                    }
+                    edgesToShapePaths.computeIfAbsent(edgeCount, s -> new TreeMap<>())
+                            .put(shape, Collections.unmodifiableList(paths));
+                }
+            }
+        }
+
+        for (Map.Entry<Integer, Map<Shape, List<PathFinder.Path>>> entry : edgesToShapePaths.entrySet()) {
+            recursiveShapes.putAll(entry.getValue());
+        }
+    }
+
+    private void visitShape(NeighborProvider provider, Shape shape) {
+        // Visit members before visiting containers. Note that no 'visited'
+        // set is needed since only non-recursive shapes are traversed.
+        for (Relationship rel : provider.getNeighbors(shape)) {
```

Comment:
Is the ordering of this stable?

Reply:
It is stable, but not alphabetically ordered. I updated it to sort neighbors alphabetically to make it more predictable.